File: cousins_binary_tree.py
# Definition for a binary tree node.
# SC : O(N) - N number of nodes
# TC : O(N) 

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

from typing import Optional
class Solution:
    def isCousins(self, root: Optional[TreeNode], x: int, y: int) -> bool:
        def parent_of_node(node, x):
            st = []
            #print("in for ", x)
            if node:
                st.append([node,0])
            if node.val == x:
                return [TreeNode(-1),-1]
            
            while st:
                parent, height = st.pop(0)
                if parent.left:
                    if parent.left.val == x:
                        return [parent,height] 
                    st.append([parent.left, height + 1])
                if parent.right:
                    if parent.right.val == x:
                        return [parent,height] 
                    st.append([parent.right, height + 1])
                    
            
            return [-1, -1] 
        
        p_x , h_x = parent_of_node(root, x)
        p_y , h_y = parent_of_node(root, y)
        
        print(p_x,h_x)
        print(p_y,h_y)
        if p_x.val != p_y.val and h_x == h_y:
            return True
        
        return False
        
        
            
        

File: cousinsbinarytree.swift
// Time Complexity : O(N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
   // Take the each row from the matrix , since the elements of the each row are in sorted order so check if the target element is present in row or not by comparing it with last element in the row, it the target is less than last element in the row then apply binary search in that row.   

    func isCousins(_ root: TreeNode?, _ x: Int, _ y: Int) -> Bool {
       if root == nil {
           return true
       } 
       var queue = [(TreeNode?,Int?)]()
        queue.append((root,root?.val))
        
        while !queue.isEmpty {
            var size = queue.count
            var parentX = -1
            var parentY = -1
            
            for i in 0..<size {
                var (node,parent) = queue.removeFirst()
                if node?.val == x {
                    parentX = parent!
                } else if node?.val == y {
                    parentY = parent!
                }
                if node?.left != nil {
                    queue.append((node?.left,node?.val))
                    
                }
                if node?.right != nil {
                    queue.append((node?.right,node?.val))
                    
                }
                
            }
            if parentX != -1 && parentY != -1 && parentX != parentY {
                return true
            }
            else if ((parentX == -1 && parentY != -1) || (parentX != -1 && parentY == -1)) {
                return false
            }
            
        } 
        return false
    }
